File: ics-runtime/ics_runtime/contracts/capability_enforcer.py
"""CapabilityEnforcer — post-execution scanner for DENY/REQUIRE violations."""

from __future__ import annotations

import re

from ics_runtime.contracts.violation import ContractViolation

# Patterns for common PII types used by deny_pii heuristics
_PII_PATTERNS = [
    (r"\b\d{3}-\d{2}-\d{4}\b", "SSN pattern"),
    (r"\b\d{4}[\s-]\d{4}[\s-]\d{4}[\s-]\d{4}\b", "credit card pattern"),
    (r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b", "email address"),
]


def _parse_directives(capability_text: str) -> list[dict]:
    """Extract ALLOW/DENY/REQUIRE lines from capability text.

    Returns a list of dicts: ``{"directive": "DENY", "rule": "..."}``.
    """
    directives: list[dict] = []
    for line in capability_text.splitlines():
        stripped = line.strip()
        for keyword in ("DENY", "REQUIRE", "ALLOW"):
            if stripped.upper().startswith(keyword):
                rule = stripped[len(keyword):].lstrip(":").strip()
                directives.append({"directive": keyword, "rule": rule})
                break
    return directives


class CapabilityEnforcer:
    """Parse DENY/REQUIRE directives and scan LLM output post-execution.

    This is the enforcement gap the ICS spec leaves to the caller.

    Enforcement is two-pass:
    1. ``on_failure`` detection — if the model itself reports a block
       (starts with ``on_failure_prefix``) we record the violation with
       ``severity="blocked"`` (highest confidence).
    2. Heuristic scanning — DENY directives containing PII/bulk/float
       keywords trigger pattern-based scans of the response text.

    Tool calls are checked via ``check_tool_call()`` before execution.
    """

    def __init__(
        self,
        capability_text: str,
        on_failure_prefix: str = "BLOCKED:",
    ) -> None:
        self._directives = _parse_directives(capability_text)
        self._on_failure_prefix = on_failure_prefix
        self._deny_rules = [d["rule"] for d in self._directives if d["directive"] == "DENY"]
        self._require_rules = [d["rule"] for d in self._directives if d["directive"] == "REQUIRE"]

    # ------------------------------------------------------------------
    # Output scanning
    # ------------------------------------------------------------------

    def scan_output(self, response_text: str) -> list[ContractViolation]:
        """Scan the final LLM response text for capability violations."""
        violations: list[ContractViolation] = []

        # Pass 1: on_failure prefix detection
        stripped = response_text.strip()
        if stripped.startswith(self._on_failure_prefix):
            rule = self._extract_blocked_rule(stripped)
            violations.append(ContractViolation(
                rule=rule or f"DENY (model self-reported block)",
                kind="capability",
                severity="blocked",
                evidence=stripped[:200],
            ))
            return violations  # Model blocked itself; skip heuristics

        # Pass 2: heuristic DENY scanning
        for rule in self._deny_rules:
            rule_lower = rule.lower()

            # PII heuristic
            if any(kw in rule_lower for kw in ("pii", "ssn", "email", "account number", "phone")):
                for pattern, label in _PII_PATTERNS:
                    m = re.search(pattern, response_text)
                    if m:
                        violations.append(ContractViolation(
                            rule=f"DENY {rule}",
                            kind="capability",
                            severity="detected",
                            evidence=m.group(0),
                        ))

            # Bulk export heuristic
            if any(kw in rule_lower for kw in ("bulk", "export", "all records")):
                if re.search(r"\ball records\b|\bexport all\b|\bCSV\b", response_text, re.I):
                    violations.append(ContractViolation(
                        rule=f"DENY {rule}",
                        kind="capability",
                        severity="detected",
                        evidence=response_text[:120],
                    ))

            # Float arithmetic on monetary values heuristic
            if "float" in rule_lower and "monetar" in rule_lower:
                if re.search(r"\bfloat\(|\.0\b.*\$|\$.*\.0\b", response_text, re.I):
                    violations.append(ContractViolation(
                        rule=f"DENY {rule}",
                        kind="capability",
                        severity="detected",
                        evidence=response_text[:120],
                    ))

        return violations

    def check_tool_call(self, tool_name: str, arguments: dict) -> list[ContractViolation]:
        """Check a proposed tool call against DENY directives.

        Called by ``Session`` before a tool is dispatched.  Returns violations
        (empty = approved to proceed).
        """
        violations: list[ContractViolation] = []
        for rule in self._deny_rules:
            rule_lower = rule.lower()
            if "bulk" in rule_lower or "export" in rule_lower:
                for k, v in arguments.items():
                    if isinstance(v, list) and len(v) > 50:
                        violations.append(ContractViolation(
                            rule=f"DENY {rule}",
                            kind="capability",
                            severity="blocked",
                            evidence=f"tool={tool_name} arg={k} count={len(v)}",
                        ))
        return violations

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def _extract_blocked_rule(self, text: str) -> str | None:
        """Try to parse the rule name from a BLOCKED: prefix response."""
        # e.g. "BLOCKED: 'DENY logging PII data' — ..."
        m = re.search(r"['\"]DENY ([^'\"]+)['\"]", text)
        if m:
            return f"DENY {m.group(1)}"
        # Fallback: everything after the prefix on the first line
        first_line = text.splitlines()[0]
        return first_line[len(self._on_failure_prefix):].strip() or None

File: ics-runtime/java/src/main/java/io/ics/runtime/contracts/CapabilityEnforcer.java
package io.ics.runtime.contracts;

import io.ics.runtime.Violation;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Post-execution scanner for DENY/REQUIRE capability violations.
 *
 * <p>Mirrors {@code ics_runtime.contracts.capability_enforcer.CapabilityEnforcer}
 * from Python with the same two-pass approach:
 * <ol>
 *   <li>on_failure prefix detection — model self-reported block → severity=blocked</li>
 *   <li>Heuristic DENY scanning — PII patterns, bulk export keywords, code-level float arithmetic</li>
 * </ol>
 */
public final class CapabilityEnforcer {

    // PII patterns
    private static final Pattern SSN_PATTERN  = Pattern.compile("\\b\\d{3}-\\d{2}-\\d{4}\\b");
    private static final Pattern CC_PATTERN   = Pattern.compile("\\b\\d{4}[\\s-]\\d{4}[\\s-]\\d{4}[\\s-]\\d{4}\\b");
    private static final Pattern EMAIL_PATTERN = Pattern.compile("[A-Za-z0-9._%+\\-]+@[A-Za-z0-9.\\-]+\\.[A-Za-z]{2,}");

    // Bulk export
    private static final Pattern BULK_PATTERN = Pattern.compile("\\ball records\\b|\\bexport all\\b|\\bCSV\\b", Pattern.CASE_INSENSITIVE);

    // Code-level float arithmetic on monetary values (NOT plain decimals in prose)
    private static final Pattern FLOAT_CODE_PATTERN = Pattern.compile(
            "\\bfloat\\s*\\(|\\bDecimal\\s*\\(\\s*str\\s*\\(" +
            "|[*/]\\s*0\\.0[0-9]+\\b.*\\$|\\$.*[*/]\\s*0\\.0[0-9]+\\b" +
            "|\\b100\\.0\\b|\\bdivide\\b.{0,30}\\$",
            Pattern.CASE_INSENSITIVE | Pattern.DOTALL);

    // Blocked-rule extractor
    private static final Pattern BLOCKED_RULE = Pattern.compile("['\"]DENY ([^'\"]+)['\"]");

    private final List<Directive> directives = new ArrayList<>();
    private final String onFailurePrefix;

    public CapabilityEnforcer(String capabilityText) {
        this(capabilityText, "BLOCKED:");
    }

    public CapabilityEnforcer(String capabilityText, String onFailurePrefix) {
        this.onFailurePrefix = onFailurePrefix;
        parseDirectives(capabilityText);
    }

    // ── Output scanning ─────────────────────────────────────────────────────

    public List<Violation> scanOutput(String responseText) {
        List<Violation> violations = new ArrayList<>();
        String stripped = responseText.strip();

        // Pass 1: on_failure prefix
        if (stripped.startsWith(onFailurePrefix)) {
            String rule = extractBlockedRule(stripped);
            violations.add(new Violation(
                    rule != null ? rule : "DENY (model self-reported block)",
                    "capability", "blocked", stripped.substring(0, Math.min(200, stripped.length()))));
            return violations;   // skip heuristics
        }

        // Pass 2: heuristic DENY scanning
        for (Directive d : directives) {
            if (!"DENY".equals(d.keyword)) continue;
            String rule = d.rule.toLowerCase();

            // PII heuristic
            if (containsAny(rule, "pii", "ssn", "email", "account number", "phone")) {
                checkPattern(SSN_PATTERN,   responseText, "DENY " + d.rule, violations);
                checkPattern(CC_PATTERN,    responseText, "DENY " + d.rule, violations);
                checkPattern(EMAIL_PATTERN, responseText, "DENY " + d.rule, violations);
            }

            // Bulk export heuristic
            if (containsAny(rule, "bulk", "export", "all records")) {
                if (BULK_PATTERN.matcher(responseText).find()) {
                    violations.add(new Violation("DENY " + d.rule, "capability", "detected",
                            responseText.substring(0, Math.min(120, responseText.length()))));
                }
            }

            // Float arithmetic on monetary values (code-level only)
            if (rule.contains("float") && rule.contains("monetar")) {
                if (FLOAT_CODE_PATTERN.matcher(responseText).find()) {
                    violations.add(new Violation("DENY " + d.rule, "capability", "detected",
                            responseText.substring(0, Math.min(120, responseText.length()))));
                }
            }
        }

        return violations;
    }

    // ── Helpers ─────────────────────────────────────────────────────────────

    private void parseDirectives(String text) {
        if (text == null || text.isBlank()) return;
        for (String line : text.split("\\r?\\n")) {
            String s = line.strip().toUpperCase();
            for (String kw : new String[]{"DENY", "REQUIRE", "ALLOW"}) {
                if (s.startsWith(kw)) {
                    String rule = line.strip().substring(kw.length()).replaceFirst("^\\s*:\\s*", "").strip();
                    directives.add(new Directive(kw, rule));
                    break;
                }
            }
        }
    }

    private static void checkPattern(Pattern p, String text, String rule, List<Violation> out) {
        Matcher m = p.matcher(text);
        if (m.find()) {
            out.add(new Violation(rule, "capability", "detected", m.group()));
        }
    }

    private static boolean containsAny(String haystack, String... needles) {
        for (String n : needles) if (haystack.contains(n)) return true;
        return false;
    }

    private String extractBlockedRule(String text) {
        Matcher m = BLOCKED_RULE.matcher(text);
        if (m.find()) return "DENY " + m.group(1);
        String first = text.split("\\r?\\n")[0];
        String after = first.substring(Math.min(onFailurePrefix.length(), first.length())).strip();
        return after.isEmpty() ? null : after;
    }

    // ── Directive record ────────────────────────────────────────────────────

    private record Directive(String keyword, String rule) {}
}
